### PR TITLE
Add exception to value alias on ExceptionInfo

### PR DESCRIPTION
**Pull request by graingert**
For easier compatibility with unittest https://github.com/pytest-dev/unittest2pytest/issues/36

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
imho this should also issue a warning so users can fix their code up

**Comment by graingert** [comment]
@RonnyPfannschmidt totally agree, what warning should it be? Deprecation warning feels fundamentally wrong here

File: src/_pytest/_code/code.py [hunk 1]
```diff
@@ -422,6 +422,8 @@ def type(self):
     def value(self):
         """the exception value"""
         return self._excinfo[1]
+    
+    exception = value
```
Comment (nicoddemus) [inline src/_pytest/_code/code.py]:
```suggestion
    @property
    def exception(self):
        msg = (
            "The '.exception' attribute is an alias to facilitate porting "
            "tests written using unittest.\n"
            "Prefer '.value' in new code."
        )  
        warnings.warn(PytestWarning(msg), stacklevel=2)
        return self.value        
```

File: changelog/5541.feature.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Add '.exception' attribute as an alias for '.value' to facilitate porting tests written using unittest.
```
Comment (nicoddemus) [inline changelog/5541.feature.rst]:
```suggestion
Add ``.exception`` attribute as an alias for ``.value`` to facilitate porting tests written using unittest.
```

**Comment by graingert** [comment]
@nicoddemus this should also apply to pytest.warns

**Comment by graingert** [comment]
@nicoddemus eg aliases for all these:

https://github.com/python/cpython/blob/3.7/Lib/unittest/case.py#L248-L261

**Comment by nicoddemus** [comment]
> @nicoddemus eg aliases for all these:

Not sure if all of those, but `warning` <=> `expected_warning` I think so. 👍

**Comment by nicoddemus** [comment]
@graingert still want to add the aliases for `pytest.warns`?

**Comment by graingert** [comment]
looking at it now

**Comment by graingert** [comment]
@nicoddemus it's actually a bit of a pain as it would involve chainging how WarningsChecker works, eg it doesn't actually store the first matching warning anywhere

**Comment by graingert** [comment]
@nicoddemus looks a bit like https://github.com/graingert/pytest/commit/4a2ced9a9d7bc09d90715de3cf959bbb60b21234

**Comment by nicoddemus** [comment]
I see. Well I think that's fine if indeed we want to add those aliases. 👍

File: src/_pytest/recwarn.py [hunk 3]
```diff
@@ -186,28 +186,33 @@ def __exit__(self, *exc_info):
         __tracebackhide__ = True
 
         # only check if we're not currently handling an exception
```
Comment (nicoddemus) [inline src/_pytest/recwarn.py]: We need a test for the `warning` attribute to avoid regressions.

File: changelog/5541.feature.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Add ``.exception`` attribute as an alias for ``.value`` to facilitate porting tests written using unittest.
```
Comment (nicoddemus) [inline changelog/5541.feature.rst]: This should be updated to mention the `warning` attribute of `pytest.warns`

File: src/_pytest/recwarn.py [hunk 5]
```diff
@@ -186,28 +186,33 @@ def __exit__(self, *exc_info):
...
+            return
+
+        if self.expected_warning is None:
+            return
+
+        for r in self:
+            if issubclass(r.category, self.expected_warning) and (
+                self.match_expr is None or re.search(self.match_expr, str(r.message))
+            ):
+                self.warning = r.message
```
Comment (nicoddemus) [inline src/_pytest/recwarn.py]: These attributes should be initialized with `None` in `__init__`

Comment (nicoddemus) [reply src/_pytest/recwarn.py]: Also, we should add a quick example to the docs to `warnings.rst`. 👍

**Comment by blueyed** [comment]
@graingert gentle ping

**Comment by graingert** [comment]
Looks like this needs a rebase etc

File: changelog/5541.feature.rst (hunk 4 above)
Comment (nicoddemus) [reply changelog/5541.feature.rst]: Oh this should also be `5541.improvement.rst` now.

**Comment by nicoddemus** [comment]
> Looks like this needs a rebase etc

Indeed, also please see my comments. 👍

File: src/_pytest/_code/code.py [hunk 6]
```diff
@@ -422,6 +422,20 @@ def type(self):
     def value(self):
         """the exception value"""
         return self._excinfo[1]
+    
+    @property
+    def exception(self):
+        """
+        an alias to '.value' to facilitate porting porting tests written using
```
Comment (blueyed) [inline src/_pytest/_code/code.py]:
```suggestion
        an alias to '.value' to facilitate porting tests written using
```

**Comment by bluetech** [comment]
Hi @graingert,

We've recently stopped using the `features` branch, and use the `master` branch instead ([rationale](https://github.com/pytest-dev/pytest/pull/6571)). We're asking open PRs against `features` to be rabased on `master`. You can do it as follows:

- `git checkout master && git pull origin` - sync the latest code.
- `git checkout patch-2` - checkout the branch of this PR.
- `git rebase -i master` - leave only your own commits, and delete everything else.
- `git push --force` - update the PR.
- Click `Edit` in the PR GitHub UI and change the base branch from `features` to `master`.

Let us know if you have any issues.

Thanks!

**Comment by nicoddemus** [comment]
TBH I'm not sure this is worthwhile doing anymore... perhaps it is better to improve [unittest2pytest](https://github.com/pytest-dev/unittest2pytest) to deal with this automatically, and mention it more prominently in the docs?

**Comment by bluetech** [comment]
> We've recently stopped using the features branch

(I've changed the base on my own now).

**Comment by bluetech** [comment]
I agree with @nicoddemus; let's close this.

**Pull request closed without merging**